### PR TITLE
Validate the provided Zendesk subdomain

### PR DESCRIPTION
**Pull request by thekindofme**
Validate the provided Zendesk subdomain to be a valid subdomain in:
- getAuthUrl
- getAccessToken

### CODE REVIEW

**Review by manted — APPROVED** [review]

File: src/Zendesk/API/Utilities/OAuth.php [hunk 1]
```diff
@@ -71,4 +74,17 @@ public static function getAuthUrl($subdomain, array $options, $domain = 'zendesk
...
+
+    /**
+     * Validate subdomain
+     *
+     * @param string $subdomain
+     * @throws InvalidArgumentException
+     */
+    private static function validateSubdomain($subdomain)
+    {
+        if (! preg_match('/^[0-9A-Za-z-]+$/', $subdomain)) {
```
Comment (SmeegeSec) [inline src/Zendesk/API/Utilities/OAuth.php:86]: Do we allow subdomains which have a preceding or trailing dash ('-')? On zendesk.com/register the subdomain validator would not accept "-subdomain" or "subdomain-" but "sub-domain" is fine. Might be worth making the match more restrictive.

Comment (thekindofme) [reply src/Zendesk/API/Utilities/OAuth.php]: @SmeegeSec make subdomain validation a bit more restrictive. Ref: 60e0589

**Review by SmeegeSec — APPROVED** [review]

**Review by veera — APPROVED** [review]

**Comment by BlackPanther73** [comment]
Add a reference to the PR comment
* https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2021-30492
* (in progress) - add the link to the security advisory

**Comment by dmitryokov** [comment]
@thekindofme how about updating the meta in composer repo? There is still an old 2.2.10 version and cannot be updated.
https://packagist.org/packages/zendesk/zendesk_api_client_php

**Comment by Kynth** [comment]
Seconded - could you update Packagist with the 2.2.11 release for Composer to pick up please?

**Comment by sanmai** [comment]
Hey folks, if you would like to review a backport of this fix to the previous version in #477, please be my guest. Thanks!

**Pull request merged**